File: legacykit-ui/src/lib/api/restore.ts
import { invoke } from '@tauri-apps/api/core';
import type { DeviceInfo } from '$lib/stores/deviceStore.svelte';

export type RestoreOptionKind =
  | 'otaDowngrade'
  | 'powdersnow'
  | 'latest'
  | 'blobRestore'
  | 'tethered'
  | 'customIpsw'
  | 'dfuIpsw'
  | 'setNonce'
  | 'ipswDownloader'
  | 'moreVersions';

export interface RestoreOption {
  kind: RestoreOptionKind;
  title: string;
  description: string;
  targetVersion: string | null;
  requiresBlobs: boolean;
  requiresDfu: boolean;
}

export interface RestoreOptionsResponse {
  productType: string | null;
  processorGeneration: number | null;
  options: RestoreOption[];
  warnings: string[];
}

export interface IpswDownloadRequest {
  url: string;
  outputDir: string;
  fileName: string | null;
}

export interface IpswDownloadResult {
  path: string;
}

export interface IpswVerifyRequest {
  path: string;
  expectedSha1: string | null;
}

export interface IpswVerifyResult {
  path: string;
  calculatedSha1: string;
  expectedSha1: string | null;
  matches: boolean | null;
}

export type RestoreTool = 'ideviceRestore' | 'futureRestore';

export interface RestoreRunRequest {
  tool: RestoreTool;
  ipswPath: string;
  shshPath: string | null;
  erase: boolean;
  update: boolean;
  usePwndfu: boolean;
  skipBlob: boolean;
  setNonce: boolean;
  noBaseband: boolean;
  latestSep: boolean;
  latestBaseband: boolean;
  dryRun: boolean;
}

export interface RestoreCommandPreview {
  supported: boolean;
  tool: RestoreTool;
  binary: string;
  args: string[];
  warnings: string[];
}

export function getRestoreOptions(device: DeviceInfo): Promise<RestoreOptionsResponse> {
  return invoke<RestoreOptionsResponse>('get_restore_options', { device });
}

export function downloadIpsw(request: IpswDownloadRequest): Promise<IpswDownloadResult> {
  return invoke<IpswDownloadResult>('download_ipsw', { request });
}

export function verifyIpsw(request: IpswVerifyRequest): Promise<IpswVerifyResult> {
  return invoke<IpswVerifyResult>('verify_ipsw', { request });
}

export function previewRestoreCommand(request: RestoreRunRequest): Promise<RestoreCommandPreview> {
  return invoke<RestoreCommandPreview>('preview_restore_command', { request });
}

export function startRestore(request: RestoreRunRequest): Promise<RestoreCommandPreview> {
  return invoke<RestoreCommandPreview>('start_restore', { request });
}

File: legacykit-ui/src-tauri/src/models/restore.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum RestoreOptionKind {
    OtaDowngrade,
    Powdersnow,
    Latest,
    BlobRestore,
    Tethered,
    CustomIpsw,
    DfuIpsw,
    SetNonce,
    IpswDownloader,
    MoreVersions,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RestoreOption {
    pub kind: RestoreOptionKind,
    pub title: String,
    pub description: String,
    pub target_version: Option<String>,
    pub requires_blobs: bool,
    pub requires_dfu: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RestoreOptionsResponse {
    pub product_type: Option<String>,
    pub processor_generation: Option<u8>,
    pub options: Vec<RestoreOption>,
    pub warnings: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct IpswDownloadRequest {
    pub url: String,
    pub output_dir: String,
    pub file_name: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct IpswDownloadResult {
    pub path: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct IpswVerifyRequest {
    pub path: String,
    pub expected_sha1: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct IpswVerifyResult {
    pub path: String,
    pub calculated_sha1: String,
    pub expected_sha1: Option<String>,
    pub matches: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum RestoreTool {
    IdeviceRestore,
    FutureRestore,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RestoreRunRequest {
    pub tool: RestoreTool,
    pub ipsw_path: String,
    pub shsh_path: Option<String>,
    pub erase: bool,
    pub update: bool,
    pub use_pwndfu: bool,
    pub skip_blob: bool,
    pub set_nonce: bool,
    pub no_baseband: bool,
    pub latest_sep: bool,
    pub latest_baseband: bool,
    pub dry_run: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RestoreCommandPreview {
    pub supported: bool,
    pub tool: RestoreTool,
    pub binary: String,
    pub args: Vec<String>,
    pub warnings: Vec<String>,
}
